### ISSUE TITLE
在训练集上效果很好，在新数据集上测试推理完全没有效果，输出全为幻觉

### ISSUE THREAD

**Issue by wangjiannan98** [Gmgge/TrOCR-Seal-Recognition#48]
您好，请教一下，我在您提供的数据集上训练了100个epochs，训练后的模型在训练集上效果很好，但是在新的印章图像上推理，却完全没有效果。训练参数没有做变动，数据差不多15000张。

这是在训练集上的推理效果
![下载](https://github.com/user-attachments/assets/4b5c1063-c541-4fd8-8e12-b67dc6b3a665)
这是在新数据上的推理效果
![下载 (1)](https://github.com/user-attachments/assets/161f4123-4fa4-4780-b449-c5491430b5bc)

**Comment by Gmgge** [comment]
你是否配置了验证集？从现象上面看是过拟合了。从我先前的经验来说通常几十个epoch就一键收敛了，你可以检查下你的训练日志，看下loss变化和验证集准确率变化。

**Comment by wangjiannan98** [comment]
> 你是否配置了验证集？从现象上面看是过拟合了。从我先前的经验来说通常几十个epoch就一键收敛了，你可以检查下你的训练日志，看下loss变化和验证集准确率变化。

谢谢您百忙中的回复，我确实没有设置验证集，我键入的命令如下：
![image](https://github.com/user-attachments/assets/913ab037-9808-4930-a7a6-e88dad10bcb1)
由于之前没有做过 transforsmer 相关的项目，想请教您，我该如何设置验证集，似乎命令中只传入了 seal_dataset一个数据集。下面是我的训练日志，似乎缺失了很多东西：
{
  "best_metric": null,
  "best_model_checkpoint": null,
  "epoch": 100.0,
  "eval_steps": 1000,
  "global_step": 143600,
  "is_hyper_param_search": false,
  "is_local_process_zero": true,
  "is_world_process_zero": true,
  "log_history": [
    {
      "epoch": 0.006963788300835654,
      "grad_norm": 1.2250148057937622,
      "learning_rate": 0.00039997493036211703,
      "loss": 2.9131,
      "step": 10
    },
    {
      "epoch": 0.013927576601671309,
      "grad_norm": 6.120301723480225,
      "learning_rate": 0.00039994707520891364,
      "loss": 1.8774,
      "step": 20
    },
    {
      "epoch": 0.020891364902506964,
      "grad_norm": 3.285402774810791,
      "learning_rate": 0.00039992200557103065,
      "loss": 1.6345,
      "step": 30
    },
    {
      "epoch": 0.027855153203342618,
      "grad_norm": 4.458782196044922,
      "learning_rate": 0.0003998941504178273,
      "loss": 1.4722,
      "step": 40
    },
    {
      "epoch": 0.034818941504178275,
      "grad_norm": 4.0512847900390625,
      "learning_rate": 0.00039986629526462397,
      "loss": 1.9359,
      "step": 50
    },
    {
      "epoch": 0.04178272980501393,
      "grad_norm": 3.1800527572631836,
      "learning_rate": 0.00039983844011142063,
      "loss": 1.1103,
      "step": 60
    },
    {
      "epoch": 0.04874651810584958,
      "grad_norm": 2.345869302749634,
      "learning_rate": 0.0003998105849582173,
      "loss": 0.8589,
      "step": 70
    },
    {
      "epoch": 0.055710306406685235,
      "grad_norm": 5.168165683746338,
      "learning_rate": 0.00039978272980501395,
      "loss": 0.8648,
      "step": 80
    },
    {
      "epoch": 0.06267409470752089,
      "grad_norm": 3.8407413959503174,
      "learning_rate": 0.00039975766016713096,
      "loss": 1.0267,
      "step": 90
    },
    {
      "epoch": 0.06963788300835655,
      "grad_norm": 3.999751567840576,
      "learning_rate": 0.00039972980501392757,
      "loss": 0.9707,
      "step": 100
    },
    {
      "epoch": 0.0766016713091922,
      "grad_norm": 1.7726722955703735,
      "learning_rate": 0.0003997019498607243,
      "loss": 0.7751,
      "step": 110
    },

     .......(中间部分省略)

 "logging_steps": 10,
  "max_steps": 143600,
  "num_input_tokens_seen": 0,
  "num_train_epochs": 100,
  "save_steps": 1000,
  "stateful_callbacks": {
    "TrainerControl": {
      "args": {
        "should_epoch_stop": false,
        "should_evaluate": false,
        "should_log": false,
        "should_save": true,
        "should_training_stop": true
      },
      "attributes": {}
    }
  },
  "total_flos": 8.983813128286372e+19,
  "train_batch_size": 8,
  "trial_name": null,
  "trial_params": null
}

    
烦请您有空的时候指教一下，十分感谢您。

**Comment by wang-zhix** [comment]
@wangjiannan98 
不需要主动设置验证集，会直接从你传入的seal_dataset中，切分5%当做验证集。
你需要在你的log日志里查找一下，类似下面的数据。来查看验证集的loss变化和准确率变化
```
        {
      "epoch": 2.49,
      "eval_acc": 0.8970499999551474,
      "eval_cer": 0.01108147774942224,
      "eval_loss": 0.026656482368707657,
      "eval_runtime": 2664.5219,
      "eval_samples_per_second": 7.506,
      "eval_steps_per_second": 0.938,
      "step": 20000
    }
```

**Comment by wangjiannan98** [comment]
> @wangjiannan98 不需要主动设置验证集，会直接从你传入的seal_dataset中，切分5%当做验证集。 你需要在你的log日志里查找一下，类似下面的数据。来查看验证集的loss变化和准确率变化
> 
> ```
>         {
>       "epoch": 2.49,
>       "eval_acc": 0.8970499999551474,
>       "eval_cer": 0.01108147774942224,
>       "eval_loss": 0.026656482368707657,
>       "eval_runtime": 2664.5219,
>       "eval_samples_per_second": 7.506,
>       "eval_steps_per_second": 0.938,
>       "step": 20000
>     }
> ```
![image](https://github.com/user-attachments/assets/e26ffd42-9464-43a8-87b2-76f87a980d85)
谢谢您的解答，从我的训练日志来看，似乎好像并没有发生过拟合，那是否就是因为我的数据量或者数据分布不够，我目前的数据集构成是飞桨的社区印章数据集（1w张）和 repo作者提供的5000张真实数据，共1.5w张图片。飞桨的社区印章数据集似乎非常的干净，不知道您是够可以提供一些优化思路，谢谢您。
![image](https://github.com/user-attachments/assets/5711e742-875c-4578-bcf4-6483ae617e6e)

**Comment by wang-zhix** [comment]
@wangjiannan98 
我认为是数据量的问题。1.5w的量训练文本行识别都不够用。
去看看假印章生成吧。
`https://github.com/Gmgge/TrOCR-Seal-Recognition/issues/38`
先用大量假样本及公开数据集进行base模型的训练，最后用与业务强相关的数据集Fine-tuning一下。
如果想要训练通用的模型，那就所有数据集合并一起训练就行了。

**Comment by wangjiannan98** [comment]
> @wangjiannan98 我认为是数据量的问题。1.5w的量训练文本行识别都不够用。 去看看假印章生成吧。 `https://github.com/Gmgge/TrOCR-Seal-Recognition/issues/38` 先用大量假样本及公开数据集进行base模型的训练，最后用与业务强相关的数据集Fine-tuning一下。 如果想要训练通用的模型，那就所有数据集合并一起训练就行了。

好的，非常感谢您的解答。

**Comment by dbzqyyy** [comment]
你好，请问一下实际应用是先检测印章再识别吧，所以还需要训练一个检测印章模型吗

**Comment by wangjiannan98** [comment]
> 你好，请问一下实际应用是先检测印章再识别吧，所以还需要训练一个检测印章模型吗

是的，需要先将印章区域抠出来再送入识别模型，检测印章模型可以看下PaddleDetection的

**Comment by Vio1etovo** [comment]
求助，目前我也使用了飞桨的社区印章数据集（1w张）和 repo作者提供的5000张真实数据，共1.5w张图片。但loss完全降不下去，而且在训练集上推理也都是幻觉。
{
  "best_metric": null,
  "best_model_checkpoint": null,
  "epoch": 3.7453183520599254,
  "eval_steps": 1000,
  "global_step": 1000,
  "is_hyper_param_search": false,
  "is_local_process_zero": true,
  "is_world_process_zero": true,
  "log_history": [
    {
      "epoch": 0.03745318352059925,
      "grad_norm": 9.326410293579102,
      "learning_rate": 0.0003988014981273408,
      "loss": 8.7464,
      "step": 10
    },
    {
      "epoch": 0.0749063670411985,
      "grad_norm": 7.00970983505249,
      "learning_rate": 0.0003973033707865169,
      "loss": 5.3504,
      "step": 20
    },
    {
      "epoch": 0.11235955056179775,
      "grad_norm": 7.767610549926758,
      "learning_rate": 0.0003958052434456929,
      "loss": 4.8916,
      "step": 30
    },
    {
      "epoch": 0.149812734082397,
      "grad_norm": 5.386384010314941,
      "learning_rate": 0.00039430711610486893,
      "loss": 4.5472,
      "step": 40
    },
    {
      "epoch": 0.18726591760299627,
      "grad_norm": 5.103940010070801,
      "learning_rate": 0.00039280898876404495,
      "loss": 4.4347,
      "step": 50
    },

----

{
      "epoch": 3.7078651685393256,
      "grad_norm": 4.392827987670898,
      "learning_rate": 0.0002519850187265918,
      "loss": 1.9744,
      "step": 990
    },
    {
      "epoch": 3.7453183520599254,
      "grad_norm": 5.080997467041016,
      "learning_rate": 0.00025048689138576784,
      "loss": 1.8701,
      "step": 1000
    },
    {
      "epoch": 3.7453183520599254,
      "eval_acc": 0.0,
      "eval_cer": 1.0390625,
      "eval_loss": 2.7817234992980957,
      "eval_runtime": 12.2455,
      "eval_samples_per_second": 9.228,
      "eval_steps_per_second": 1.225,
      "step": 1000
    }
  ],

**Comment by wang-zhix** [comment]
@Vio1etovo 
你贴的这个日志，loss不是一直在降吗？

**Comment by wangjiannan98** [comment]
> 求助，目前我也使用了飞桨的社区印章数据集（1w张）和 repo作者提供的5000张真实数据，共1.5w张图片。但loss完全降不下去，而且在训练集上推理也都是幻觉。 { "best_metric": null, "best_model_checkpoint": null, "epoch": 3.7453183520599254, "eval_steps": 1000, "global_step": 1000, "is_hyper_param_search": false, "is_local_process_zero": true, "is_world_process_zero": true, "log_history": [ { "epoch": 0.03745318352059925, "grad_norm": 9.326410293579102, "learning_rate": 0.0003988014981273408, "loss": 8.7464, "step": 10 }, { "epoch": 0.0749063670411985, "grad_norm": 7.00970983505249, "learning_rate": 0.0003973033707865169, "loss": 5.3504, "step": 20 }, { "epoch": 0.11235955056179775, "grad_norm": 7.767610549926758, "learning_rate": 0.0003958052434456929, "loss": 4.8916, "step": 30 }, { "epoch": 0.149812734082397, "grad_norm": 5.386384010314941, "learning_rate": 0.00039430711610486893, "loss": 4.5472, "step": 40 }, { "epoch": 0.18726591760299627, "grad_norm": 5.103940010070801, "learning_rate": 0.00039280898876404495, "loss": 4.4347, "step": 50 },
> 
> { "epoch": 3.7078651685393256, "grad_norm": 4.392827987670898, "learning_rate": 0.0002519850187265918, "loss": 1.9744, "step": 990 }, { "epoch": 3.7453183520599254, "grad_norm": 5.080997467041016, "learning_rate": 0.00025048689138576784, "loss": 1.8701, "step": 1000 }, { "epoch": 3.7453183520599254, "eval_acc": 0.0, "eval_cer": 1.0390625, "eval_loss": 2.7817234992980957, "eval_runtime": 12.2455, "eval_samples_per_second": 9.228, "eval_steps_per_second": 1.225, "step": 1000 } ],

看你的日志，loss是在下降的，而且训练不到4个epoch，模型还没有充分拟合训练集

**Comment by wang-zhix** [comment]
@Vio1etovo 
作者有提供预训练模型，可以尝试一下。

**Comment by Vio1etovo** [comment]
> @Vio1etovo 作者有提供预训练模型，可以尝试一下。

使用了作者提供的weights 、TAL_OCR_CHN预训练模型和1.5w数据集，在预设参数下训练10个epoch，valset acc也很低。
{
      "epoch": 9.25925925925926,
      "grad_norm": 3.3181796073913574,
      "learning_rate": 2.9797979797979796e-05,
      "loss": 0.516,
      "step": 11000
    },
    {
      "epoch": 9.25925925925926,
      "eval_acc": 0.025999999948,
      "eval_cer": 0.6601331637615022,
      "eval_loss": 0.4298277497291565,
      "eval_runtime": 71.731,
      "eval_samples_per_second": 6.97,
      "eval_steps_per_second": 0.878,
      "step": 11000
    },

**Comment by Vio1etovo** [comment]
此外，我还爬取了40w+的企业名称，并利用假印章生成工具来生成模拟数据，结合作者提供1.5w数据和我收集的其他数据作为训练集，目前正在训练中。
{'eval_loss': 2.6163206100463867, 'eval_cer': 0.8285195634076403, 'eval_acc': 0.0, 'eval_runtime': 555.724, 'eval_samples_per_second': 7.047, 'eval_steps_per_second': 0.882, 'epoch': 0.22}

**Comment by wang-zhix** [comment]
@Vio1etovo 
目前看到的问题只有训练时长不够。因为loss还是在降的。

**Comment by wangjiannan98** [comment]
> @wangjiannan98 我认为是数据量的问题。1.5w的量训练文本行识别都不够用。 去看看假印章生成吧。 `https://github.com/Gmgge/TrOCR-Seal-Recognition/issues/38` 先用大量假样本及公开数据集进行base模型的训练，最后用与业务强相关的数据集Fine-tuning一下。 如果想要训练通用的模型，那就所有数据集合并一起训练就行了。

您好，我用您提供的假印章生成脚本生成了10w张数据，一共12w张图片在单卡上进行训练，启动训练脚本后，一直卡在 loading data 这里，请教您一下这是什么原因。

**Comment by wang-zhix** [comment]
@wangjiannan98 
之前能跑，加完数据就不行了。 那就是加的数据格式有问题。
生成的数据格式不一定符合代码加载的格式，你得修改同步一下。

**Comment by wangjiannan98** [comment]
> @wangjiannan98 生成的数据格式不一定符合代码加载的格式，你得修改同步一下。
> 
> 代码支持的是这样的，一个jpg 对应一个txt 0.jpg # 图像 0.txt # 标签 ... 100000.jpg 100000.txt
> 
> 生成的样本 只有一个txt，保存了所有标签。

数据我像您说的这样处理过的，启动训练后没有报错，一直卡在loading data 这里（一个小时都没有变化）
![image](https://github.com/user-attachments/assets/4ca9d6e1-f5dc-49d1-a280-b17a9f8b26ac)
GPU 监测一直没有出现显存占用，我怀疑是cpu限制了，但是CPU也没有出现很高的占用，batch_size我设置的 8 ， 也并不大。请教您一下，12w数据训练Transformer 大概需要什么样的硬件配置。

**Comment by wangjiannan98** [comment]
> @wangjiannan98 加载数据不会吃硬件的，别说12W 就是120W 也是能加载的。（加载的时候，只有文件路径，不会直接读图，所以不需要考虑硬件问题） 看一下get_image_file_list函数 我怀疑是它在遍历12W数据，加上txt，实际要遍历24W数据。 你可以单独运行一下tool/file_tool.py，确认一下是不是它引起的。
> 
> 如果确定是这里，你可以修改代码，直接读取生成数据那种格式的txt。

好的，十分感谢您

**Comment by Adorablepet** [comment]
> 此外，我还爬取了40w+的企业名称，并利用假印章生成工具来生成模拟数据，结合作者提供1.5w数据和我收集的其他数据作为训练集，目前正在训练中。 {'eval_loss': 2.6163206100463867, 'eval_cer': 0.8285195634076403, 'eval_acc': 0.0, 'eval_runtime': 555.724, 'eval_samples_per_second': 7.047, 'eval_steps_per_second': 0.882, 'epoch': 0.22}

@wangjiannan98 兄弟，爬取这个40w+的企业名称能共享下吗？你在哪个网站爬的呢？谢谢
